block-producer/src/server/api.rs

```diff
@@ -21,38 +32,138 @@ use crate::{
+                    account_hash: Some(account_hash.into()),
+                })
+                .collect(),
+            nullifiers: block
+                .produced_nullifiers
+                .iter()
+                .map(|nullifier| nullifier.into())
+                .collect(),
+            notes: todo!(),
+        });
+
+        let _ = self
+            .store
+            .apply_block(request)
+            .await
+            .map_err(|status| ApplyBlockError::GrpcClientError(status.message().to_string()))?;
+
+        Ok(())
     }
 }
 
 #[async_trait]
-impl Store for RpcStore {
+impl Store for DefaultStore {
     async fn get_tx_inputs(
         &self,
-        _proven_tx: SharedProvenTx,
+        proven_tx: SharedProvenTx,
     ) -> Result<TxInputs, TxInputsError> {
-        todo!()
+        let request = tonic::Request::new(GetTransactionInputsRequest {
+            account_ids: vec![proven_tx.account_id().into()],
+            nullifiers: proven_tx
+                .consumed_notes()
+                .iter()
+                .map(|note| note.nullifier().into())
+                .collect(),
+        });
+        let response = self
+            .store
+            .clone()
+            .get_transaction_inputs(request)
+            .await
+            .map_err(|status| TxInputsError::GrpcClientError(status.message().to_string()))?
+            .into_inner();
+
+        let account_hash = {
+            let account_state = response
+                .account_states
+                .get(0)
+                .ok_or(TxInputsError::MalformedResponse("account_states empty".to_string()))?;
+
+            let account_id_from_store: AccountId = account_state
+                .account_id
+                .clone()
+                .ok_or(TxInputsError::MalformedResponse("empty account id".to_string()))?
+                .try_into()?;
+
+            if account_id_from_store != proven_tx.account_id() {
+                return Err(TxInputsError::MalformedResponse(format!(
+                    "incorrect account id returned from store. Got: {}, expected: {}",
+                    account_id_from_store,
+                    proven_tx.account_id()
+                )));
+            }
+
+            account_state.account_hash.clone().map(Digest::try_from).transpose()?
+        };
+
+        let nullifiers = {
+            let mut nullifiers = Vec::new();
+
+            for nullifier_record in response.nullifiers {
+                let nullifier = nullifier_record
+                    .nullifier
+                    .ok_or(TxInputsError::MalformedResponse(
+                        "nullifier record contains empty nullifier".to_string(),
+                    ))?
+                    .try_into()?;
+
+                // `block_num` is nonzero if already consumed; 0 otherwise
+                nullifiers.push((nullifier, nullifier_record.block_num != 0))
+            }
+
+            nullifiers.into_iter().collect()
+        };
```

Comment:
maybe this can be replaced with [`try_convert`](https://github.com/0xPolygonMiden/miden-node/blob/main/proto/src/conversion.rs#L340-L346)?

Reply:
I cannot use `try_convert` because I cannot implement `TryFrom<NullifierTransactionInputRecord> for (RpoDigest, bool)` where `Error = TxInputsError`, since `TxInputsError` lives in `miden-node-block-producer` (and our conversion code lives in `miden-node-proto`).

Related: https://github.com/0xPolygonMiden/miden-node/issues/74#issuecomment-1850782605